sdk/core/azure-core/test/ut/url.cpp

```diff
@@ -0,0 +1,226 @@
+    EXPECT_PRED2(
+        [](std::string a, std::string b) { return a == b; },
+        url.GetAbsoluteUrl(),
+        "http://test.com?query=va=lue");
+
+    // Provide more than one extra symbol to be encoded
+    EXPECT_NO_THROW(url.AppendQueryParameter("query", Http::Url::Encode("va=l u?e", " ?")));
+    EXPECT_PRED2(
+        [](std::string a, std::string b) { return a == b; },
+        url.GetAbsoluteUrl(),
+        "http://test.com?query=va%3Dl u?e");
+
+    // default, encode it all
+    EXPECT_NO_THROW(url.AppendQueryParameter("query", Http::Url::Encode("va=l u?e")));
+    EXPECT_PRED2(
+        [](std::string a, std::string b) { return a == b; },
+        url.GetAbsoluteUrl(),
+        "http://test.com?query=va%3Dl%20u%3Fe");
+  }
+
+  TEST(URL, add_path)
+  {
+    Http::HttpMethod httpMethod = Http::HttpMethod::Post;
+    Http::Url url("http://test.com");
+    Http::Request req(httpMethod, url);
+
+    EXPECT_NO_THROW(req.GetUrl().AppendPath("path"));
+    EXPECT_PRED2(
+        [](std::string a, std::string b) { return a == b; },
+        req.GetUrl().GetAbsoluteUrl(),
+        "http://test.com/path");
+
+    EXPECT_NO_THROW(req.GetUrl().AppendQueryParameter("query", "value"));
+    EXPECT_PRED2(
+        [](std::string a, std::string b) { return a == b; },
+        req.GetUrl().GetAbsoluteUrl(),
+        "http://test.com/path?query=value");
+
+    EXPECT_NO_THROW(req.GetUrl().AppendPath("path2"));
+    EXPECT_PRED2(
+        [](std::string a, std::string b) { return a == b; },
+        req.GetUrl().GetAbsoluteUrl(),
+        "http://test.com/path/path2?query=value");
+
+    EXPECT_NO_THROW(req.GetUrl().AppendPath("path3"));
+    EXPECT_PRED2(
+        [](std::string a, std::string b) { return a == b; },
+        req.GetUrl().GetAbsoluteUrl(),
+        "http://test.com/path/path2/path3?query=value");
+  }
+
+  TEST(URL, getPort)
+  {
+    Http::Url url("http://test.com:9090");
+    int expected = 9090;
+
+    EXPECT_PRED2(
+        [](int expectedValue, int value) { return expectedValue == value; },
+        url.GetPort(),
+        expected);
+  }
+
+  TEST(URL, getPortConst)
+  {
+    Http::Url const url("https://test.com:500");
+    int expected = 500;
+
+    EXPECT_PRED2(
+        [](int expectedValue, int code) { return expectedValue == code; }, url.GetPort(), expected);
+  }
+
+  TEST(URL, getPortAfterSet)
+  {
+    Http::Url url("http://test.com");
+    url.SetPort(40);
+    int expected = 40;
+
+    EXPECT_PRED2(
+        [](int expectedValue, int code) { return expectedValue == code; }, url.GetPort(), expected);
+
+    url.SetPort(90);
+    expected = 90;
+
+    EXPECT_PRED2(
+        [](int expectedValue, int code) { return expectedValue == code; }, url.GetPort(), expected);
+  }
+
+  TEST(URL, getPortInvalidArg)
+  {
+    EXPECT_THROW(Http::Url url("http://test.com:ThisIsNotAPort"), std::invalid_argument);
+  }
+
+  TEST(URL, getPortOutfRange)
```

Comment:
Valid ports are 0-65535.  Add a test case to check the boundary condition, 65536 is an invalid port

Reply:
I think we don't have any validation for that or negative numbers. 
Following the SDK rules of "do not validate input data for the request", should we avoid preventing someone to add any number (including negative) for the port?

Reply:
I think we should try to follow at least some "domain specific" constraints and validation.